tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+  This can be described as `data V1` (note: it has no constructors).
+
+* Constructors without arguments. Let this be something like this: `data U1
+  = U1`.
+
+* Sums: `data (f :+: g) = L1 f | R1 g`.
+
+* Products: `data (f :*: g) = f :*: g`.
+
+Not bad for a start. Let's try use this representation to derive a `Functor`
+instance. Deriving a such an instance means that we should provide `fmap`
+function that looks like this:
+
+```haskell
+fmap :: (a -> b) -> Rep f -> Rep f
+```
+
+…where `Rep f` maps to the type of generic representation of `f :: * -> *`.
+There is a problem though. You see, the functors' inner type that goes from
+`a` to `b` in this example has no place to live in the representation! Our
+current approach will work only for type classes that work with things of
+kind `*`, such as `Show`, but if we want to use this system to derive a
+`Functor` instance, we need to allow it to work with kind `* -> *`.
+
+We want to be able to derive `Functor` (and similar) instances generically,
+so we have to fix things up. The solution is to add one more type parameter
+`p` to all our combinators:
+
+```haskell
+data V1 p
+data U1 p = U1
+data (f :+: g) p = L1 (f p) | R1 (g p)
+data (f :*: g) p = (f p) :*: (g p)
+```
+
+This way `fmap` would be:
+
+```haskell
+fmap :: (a -> b) -> Rep f a -> Rep f b
+--                  p  =  a    p  =  b
+```
+
+But what happens to the type classes that work with `*` kinds? Here are our
+choices:
+
+* Have a separate set of combinator types for each case (`*` and `* -> *`
+  kinds).
+
+* Use the most general form (with `p`), but for `*` kinds just treat the
+  extra `p` parameter as a dummy type index that has no meaning.
+
+The authors of the generic extension went with the second option, and I
+can't blame them. We'll see that there are a lot of wrappers already and we
+really should try to keep their number from exploding any further.
+
+Let's try to map from a data type to its representation and see if we're
+still missing something:
+
+```haskell
+data Maybe a = Nothing | Just a
+
+-- Interestingly, we could build a representation that works on ‘Maybe a’,
+-- that is, a thing of kind *, if we wanted to derive something like ‘Show’.
+-- At the same time if we wanted to derive ‘Functor’, we would work with
+-- ‘Maybe’ of kind * -> *. This means that there are actually two different
+-- possible representations depending on our aim. This is addressed with two
+-- different generics type classes, as we will see later.
+
+-- For kind *, things like ‘Show’:
+
+-- type: (U1 :+: ?) p
+```
+
+Oh, how to represent `Just a`? We need to way to allow it have an argument!
+Let's add the following:
+
+```haskell
+data Rec0 c p = Rec0 { unRec0 :: c }
+```
+
+`Rec` part in the type's name hints that this thing may be possibly
+recursive.
+
+But I'm lying to you, due to some historical stuff, it's defined a bit
+differently:
+
+```haskell
+type Rec0 = K1 R
+
```

Comment:
I am a little confused as to why the `p` parameter is a dummy in this case, where we are trying to represent a type of kind `(* -> *)`.  I was under the impression from earlier that `p` should only be a dummy if we are looking at types of kind `*`.